### PR TITLE
add sidekiq-migrate

### PR DESCRIPTION
**Pull request by izakp**
Script to migrate Sidekiq queues from one Redis to another.

I sourced the idea from https://github.com/mperham/sidekiq/issues/3900#issuecomment-410104469 then refactored a bit.

My first try migrating Convection staging, all scheduled jobs were executed immediately.  So I dropped the Sidekiq dependency in the source code, and dropped down a level to migrate everything using only Redis commands.

Also refactored so that keys remain in place in the old Redis, unless CLEAN_UP is supplied.

After testing importing to my local Redis, I felt bold, and ran the import on Convection production, knowing that without setting `CLEAN_UP=true` I could retry if anything went wrong.  But it didn't!  And you can now see all retry / scheduled jobs that were on our previous Redis instance in https://convection.artsy.net/admin/sidekiq/retries

No need now for creating a "legacy" Sidekiq deployment as in https://github.com/artsy/convection/pull/416 - we can simply update the Redis URL, then run this script to move any existing jobs to the new Redis

### CODE REVIEW

File: sidekiq-migrate/migrate.rb [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+old_redis = Redis.new(url: ENV.fetch('SIDEKIQ_OLD_REDIS_URL'))
+new_redis = Redis.new(url: ENV.fetch('SIDEKIQ_NEW_REDIS_URL'))
+
+old_redis.smembers('queues').each do |q|
+  puts "Migrating queue #{q}..."
+
+  moved_jobs_counter = Hash.new(0)
+  rqn = "queue:#{q}"
+
+  if actual_run and clean_up
```
Comment (izakp) [inline sidekiq-migrate/migrate.rb]: I don't like this branching but with Redis lists, we can either get the entire list or pop an element from the end (one way or the other) - the original code popped then pushed back to the old Redis which I found risky, so although this smells a bit, it's a lot safer IMHO

**Comment by izakp** [comment]
> This depends on a lot of sidekiq internals (especially how its jobs, retries, and schedules are stored), so I worry that these assumptions might not apply across the variety of sidekiq versions and queue options that we might use across the ~16 different applications

Given its simplicity (queues are lists of JSON blobs, retries and scheduled jobs are sorted sets of JSON blob as element / timestamp of when to execute as the score) I would be surprised if this is changed across Sidekiq versions.  I can check though.

In terms of safety, this allows us to:
- Migrate _without_ removing the old jobs, even to test on a local Redis

> One thing to confirm about the migration process: for consistency, this presumes that workers are completely stopped while migrating job data, right? If workers were reading from the old or new queues, I could imagine some race conditions.

I propose to migrate without cleaning up old queues - in this case we should migrate everything to the new redis before.

If we use the migrate with clean up option, Redis' atomicity should handle race conditions, even if we had a worker listening on new and old queues

**Comment by izakp** [comment]
@joeyAghion looking more closely at https://github.com/mperham/sidekiq/wiki/API#named-queues I added the "dead" set as well - otherwise since this script doesn't touch job JSON I think it's quite safe - the data structures and their content stay the same in Redis

**Comment by joeyAghion** [comment]
👍 

I'm still confused about the combination of "I propose to migrate without cleaning up old queues..." and "If we use the migrate with clean up option...." Which approach will we take? If a worker is still reading from the old queue at migration time, I'm guessing _with clean up_ is necessary.

**Comment by izakp** [comment]
> Which approach will we take? If a worker is still reading from the old queue at migration time, I'm guessing with clean up is necessary.

True, let's just avoid having sidekiq reading from both old and new completely.  I edited my response above...

**Comment by izakp** [comment]
@joeyAghion in regards to sidekiq version - see @mperham 's [response](https://github.com/mperham/sidekiq/issues/3900#issuecomment-410119959) in the above thread - he writes a 1-liner using Redis' [MIGRATE](https://redis.io/commands/migrate) command so it seems that migrating the lists / sorted sets wholesale is tacitly approved.

I didn't take that approach though as it looks like using Redis' MIGRATE would not merge keys if existing in the target database, so I took this safer approach.  This entire tool could be simpler though, if we assume the destination Redis to be empty (couldn't figure out what the default behavior of MIGRATE is is a key exists).

**Comment by izakp** [comment]
Check this thread for the use of the MIGRATE command to move an entire database https://stackoverflow.com/questions/37166947/copying-all-keys-in-redis-database-using-migrate

**Pull request merged**